File: Data Structures & Algorithms/design-hashmap/submission-1.py
class ListNode:
    def __init__(self, key=0, value=0, next = None):
        self.key = key
        self.value = value
        self.next = next
class MyHashMap:

    def __init__(self):
        self.map = [ListNode() for i in range(10**4)]
    
    def idx(self, key):
        return key % len(self.map)

    def put(self, key: int, value: int) -> None:
        cur = self.map[self.idx(key)]
        while cur.next!= None:
            if cur.next.key == key:
                cur.next.value = value
                return
            cur = cur.next
        cur.next = ListNode(key, value)

    def get(self, key: int) -> int:
        cur = self.map[self.idx(key)].next
        while cur:
            if cur.key == key:
                return cur.value
            cur = cur.next
        return - 1

    def remove(self, key: int) -> None:
        cur = self.map[self.idx(key)]
        while cur and cur.next:
            if cur.next.key == key:
                cur.next = cur.next.next
                return
            cur = cur.next        


# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)

File: Data Structures & Algorithms/range-sum-query-2d-immutable/submission-1.java
class NumMatrix {
    private int[][] sumMat;

    public NumMatrix(int[][] matrix) {
        int ROWS = matrix.length;
        int COLS = matrix[0].length;

        sumMat = new int [ROWS + 1][COLS + 1];

        for(int r = 0; r < ROWS; r++){
            int prefix = 0;
            for(int c = 0; c < COLS; c++){
                prefix += matrix[r][c];
                int above = sumMat[r][c + 1];
                sumMat[r + 1][c + 1] = prefix + above;
            }
        }
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        int r1 = row1 + 1;
        int c1 = col1 + 1;
        int r2 = row2 + 1;
        int c2 = col2 + 1;

        int bottomRight = sumMat[r2][c2];
        int above = sumMat[r1 - 1][c2];
        int left = sumMat[r2][c1 - 1];
        int topLeft = sumMat[r1 - 1][c1 - 1];

        return bottomRight - above - left + topLeft;
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */